File: src/main/java/no/fintlabs/provider/kafka/EventTopicEnsurer.java
package no.fintlabs.provider.kafka;

import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import no.fintlabs.provider.config.ProviderProperties;
import no.novari.kafka.topic.name.EventTopicNameParameters;
import no.novari.kafka.topic.name.TopicNamePrefixParameters;
import org.springframework.stereotype.Component;

import java.util.Map;

import static no.fintlabs.provider.kafka.TopicNamesConstants.ADAPTER_DELETE_SYNC_EVENT_NAME;
import static no.fintlabs.provider.kafka.TopicNamesConstants.ADAPTER_DELTA_SYNC_EVENT_NAME;
import static no.fintlabs.provider.kafka.TopicNamesConstants.ADAPTER_FULL_SYNC_EVENT_NAME;
import static no.fintlabs.provider.kafka.TopicNamesConstants.ADAPTER_REGISTER_EVENT_NAME;
import static no.fintlabs.provider.kafka.TopicNamesConstants.HEARTBEAT_EVENT_NAME;

@Component
@RequiredArgsConstructor
public class EventTopicEnsurer {

    private final ProviderTopicService providerTopicService;
    private final ProviderProperties providerProperties;

    @PostConstruct
    public void init() {
        ensureEventTopics(providerProperties);
    }

    private void ensureEventTopics(ProviderProperties providerProperties) {
        Map.of(
                HEARTBEAT_EVENT_NAME, providerProperties.getAdapterHeartbeatRetentionTime(),
                ADAPTER_REGISTER_EVENT_NAME, providerProperties.getAdapterRegisterRetentionTime(),
                ADAPTER_FULL_SYNC_EVENT_NAME, providerProperties.getAdapterFullSyncRetentionTime(),
                ADAPTER_DELTA_SYNC_EVENT_NAME, providerProperties.getAdapterDeltaSyncRetentionTime(),
                ADAPTER_DELETE_SYNC_EVENT_NAME, providerProperties.getAdapterDeleteSyncRetentionTime()
        ).forEach((eventName, retentionTime) -> providerTopicService.createOrModifyTopic(
                EventTopicNameParameters.builder()
                        .topicNamePrefixParameters(
                                TopicNamePrefixParameters
                                        .stepBuilder()
                                        .orgIdApplicationDefault()
                                        .domainContextApplicationDefault()
                                        .build()
                        )
                        .eventName(eventName)
                        .build(),
                retentionTime
        ));
    }

}

File: src/main/java/no/fintlabs/provider/kafka/EventTopicEnsurer.kt
package no.fintlabs.provider.kafka

import no.fintlabs.provider.config.ProviderProperties
import no.novari.kafka.topic.EventTopicService
import no.novari.kafka.topic.configuration.EventCleanupFrequency
import no.novari.kafka.topic.configuration.EventTopicConfiguration
import no.novari.kafka.topic.name.EventTopicNameParameters
import no.novari.kafka.topic.name.TopicNamePrefixParameters
import org.springframework.boot.context.event.ApplicationReadyEvent
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component

@Component
class EventTopicEnsurer(
    private val providerProperties: ProviderProperties,
    private val eventTopicService: EventTopicService
) {
    @EventListener(ApplicationReadyEvent::class)
    fun ensureEventTopics() = with(providerProperties.kafka.adapter) {
        listOf(
            TopicNamesConstants.HEARTBEAT_EVENT_NAME to heartbeatRetentionTime,
            TopicNamesConstants.ADAPTER_REGISTER_EVENT_NAME to registerRetentionTime,
            TopicNamesConstants.ADAPTER_FULL_SYNC_EVENT_NAME to fullSyncRetentionTime,
            TopicNamesConstants.ADAPTER_DELTA_SYNC_EVENT_NAME to deltaSyncRetentionTime,
            TopicNamesConstants.ADAPTER_DELETE_SYNC_EVENT_NAME to deleteSyncRetentionTime
        ).forEach { (eventName, retentionTime) ->
            eventTopicService.createOrModifyTopic(
                EventTopicNameParameters.builder()
                    .topicNamePrefixParameters(
                        TopicNamePrefixParameters.stepBuilder()
                            .orgIdApplicationDefault()
                            .domainContextApplicationDefault()
                            .build()
                    )
                    .eventName(eventName)
                    .build(),
                EventTopicConfiguration
                    .stepBuilder()
                    .partitions(providerProperties.kafka.entityPartitions)
                    .retentionTime(retentionTime)
                    .cleanupFrequency(EventCleanupFrequency.NORMAL)
                    .build()
            )
        }
    }
}